cobalt/akn.py

```diff
@@ -0,0 +1,403 @@
+        *name* is a dotted path from *self*, *after* is where to place the new
+        element if it doesn't exist. """
+        node = self._get(name)
+        if node is None:
+            # TODO: what if nodes in the path don't exist?
+            node = self._make(name.split('.')[-1])
+            after.addnext(node)
+
+        return node
+
+    def _ensure_lifecycle(self):
+        try:
+            after = self.meta.publication
+        except AttributeError:
+            after = self.meta.identification
+        node = self._ensure('meta.lifecycle', after=after)
+
+        if not node.get('source'):
+            node.set('source', '#' + self.source[1])
+            self._ensure_reference('TLCOrganization', self.source[0], self.source[1], self.source[2])
+
+        return node
+
+    def _ensure_reference(self, elem, name, id, href):
+        references = self._ensure('meta.references', after=self._ensure_lifecycle())
+
+        ref = references.find('./{*}%s[@id="%s"]' % (elem, id))
+        if ref is None:
+            ref = self._make(elem)
+            ref.set('id', id)
+            ref.set('href', href)
+            ref.set('showAs', name)
+            references.insert(0, ref)
+        return ref
+
+    def _make(self, elem):
+        return getattr(self._maker, elem)()
+
+    def _get(self, name, root=None):
+        parts = name.split('.')
+        node = root or self
+
+        for p in parts:
+            try:
+                node = getattr(node, p)
+            except AttributeError:
+                return None
+        return node
+
+
+class AmendmentStructure(StructuredDocument):
+    structure_type = "amendmentStructure"
+    main_content_element = "amendmentBody"
+
+
+class CollectionStructure(StructuredDocument):
+    structure_type = "collectionStructure"
+    main_content_element = "collectionBody"
+
+
+class DebateStructure(StructuredDocument):
+    structure_type = "debateStructure"
+    main_content_element = "debateBody"
+
+
+class HierarchicalStructure(StructuredDocument):
+    structure_type = "hierarchicalStructure"
+    main_content_element = "body"
+
+
+class JudgmentStructure(StructuredDocument):
+    structure_type = "judgmentStructure"
+    main_content_element = "judgmentBody"
+
+
+class OpenStructure(StructuredDocument):
+    structure_type = "openStructure"
+    main_content_element = "mainBody"
+
+
+class PortionStructure(StructuredDocument):
+    structure_type = "portionStructure"
+    main_content_element = "portionBody"
+
+
+# ----------------------------------------------------------------------------------
+# Document type classes
+# the following have moved into their own files:
+# - act
+# - judgment
+
+
+class Amendment(AmendmentStructure):
+    document_type = "amendment"
+
+
+class AmendmentList(CollectionStructure):
+    document_type = "amendmentList"
+
+
+class Bill(HierarchicalStructure):
+    document_type = "bill"
+
+
+class Collection(CollectionStructure):
```

Comment:
This one will be fun, because we should be able to extract each document in the collection as its own top-level object. Which implies that if you're working with, say, a `Bill` object, it could be part of a bigger AKN document, so we can't assume that `root` is the root of the XML tree. We'll have to think through this.